File: assignments/assignment-2-solution/src-csharp/TeaShop/UserInterface/QueryBuilder/InventoryQueryOutput.cs
using Assignment2Solution.Domain.InventoryQuery;

namespace Assignment2Solution.UserInterface.QueryBuilder;

/// <summary>
///     Represents the result of an inventory query, ready for display.
/// </summary>
/// <param name="Items">The list of queried inventory items.</param>
/// <param name="AppliedFiltersAndSorts">The list of filters and sorts applied to the query.</param>
public sealed record InventoryQueryOutput(
    IReadOnlyList<QueriedInventoryItem> Items,
    IReadOnlyList<string> AppliedFiltersAndSorts)
{
    /// <summary>
    ///     Creates a <see cref="InventoryQueryOutput" /> from an <see cref="IInventoryQuery" />.
    /// </summary>
    /// <param name="query">The query to execute.</param>
    /// <returns>A new query output.</returns>
    /// <exception cref="ArgumentNullException">Thrown when query is null.</exception>
    public static InventoryQueryOutput From(IInventoryQuery query)
    {
        if (query is null) throw new ArgumentNullException(nameof(query));

        // query.Execute() already returns an IReadOnlyList, 
        // but we still need to wrap them into QueriedInventoryItem with index.
        var items = query.Execute()
            .Select((item, index) => new QueriedInventoryItem(
                index + 1,
                item))
            .ToList()
            .AsReadOnly();
        var applied = query.AppliedFiltersAndSorts.ToList().AsReadOnly();

        return new InventoryQueryOutput(items, applied);
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/userinterface/querybuilder/InventoryQueryOutput.java
package assignment2solution.userinterface.querybuilder;

import assignment2solution.domain.inventoryquery.IInventoryQuery;
import assignment2solution.domain.inventoryquery.QueriedInventoryItem;

import java.util.List;
import java.util.Objects;

/**
 * Represents the result of an inventory query, ready for display.
 */
public record InventoryQueryOutput(
    List<QueriedInventoryItem> items,
    List<String> appliedFiltersAndSorts
) {
    /**
     * Creates an {@link InventoryQueryOutput} from an {@link IInventoryQuery}.
     */
    public static InventoryQueryOutput from(IInventoryQuery query) {
        Objects.requireNonNull(query, "query");

        var results = query.execute();
        var indexed = new java.util.ArrayList<QueriedInventoryItem>(results.size());
        for (int i = 0; i < results.size(); i++) {
            indexed.add(new QueriedInventoryItem(i + 1, results.get(i)));
        }

        return new InventoryQueryOutput(List.copyOf(indexed), List.copyOf(query.getAppliedFiltersAndSorts()));
    }
}
